pkg/server/container_remove.go

```diff
@@ -17,14 +17,86 @@ limitations under the License.
 package server
 
 import (
-	"errors"
+	"fmt"
 
+	"github.com/golang/glog"
 	"golang.org/x/net/context"
 
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
 )
 
 // RemoveContainer removes the container.
-func (c *criContainerdService) RemoveContainer(ctx context.Context, r *runtime.RemoveContainerRequest) (*runtime.RemoveContainerResponse, error) {
-	return nil, errors.New("not implemented")
+func (c *criContainerdService) RemoveContainer(ctx context.Context, r *runtime.RemoveContainerRequest) (retRes *runtime.RemoveContainerResponse, retErr error) {
+	glog.V(2).Infof("RemoveContainer for %q", r.GetContainerId())
+	defer func() {
+		if retErr == nil {
+			glog.V(2).Infof("RemoveContainer returns successfully")
+		}
+	}()
+
+	id := r.GetContainerId()
+
+	// Set removing state to prevent following start and removing.
+	err := c.setContainerRemoving(id)
+	if err != nil {
+		if !metadata.IsNotExistError(err) {
+			return nil, fmt.Errorf("failed to set removing state for container %q: %v",
+				id, err)
+		}
+		// Do not return error if container metadata doesn't exist.
+		glog.V(5).Infof("RemoveContainer called for container %q that does not exist", id)
+		return &runtime.RemoveContainerResponse{}, nil
+	}
+	defer func() {
+		if retErr == nil {
+			// Cleanup all index after successfully remove the container.
+			c.containerNameIndex.ReleaseByKey(id)
+			return
+		}
+		// Reset removing if remove failed.
+		err := c.containerStore.Update(id, func(meta metadata.ContainerMetadata) (metadata.ContainerMetadata, error) {
+			meta.Removing = false
+			return meta, nil
+		})
+		// TODO(random-liu): Deal with update failure. Actually Removing doesn't need to
+		// be checkpointed, we only need it to have the same lifecycle with container metadata.
+		if err != nil {
+			glog.Errorf("failed to reset removing state for container %q: %v",
+				id, err)
+		}
+	}()
+
+	// TODO(random-liu): [P0] Cleanup container rootfs.
+
+	// Cleanup container root directory.
+	containerRootDir := getContainerRootDir(c.rootDir, id)
+	if err := c.os.RemoveAll(containerRootDir); err != nil {
+		return nil, fmt.Errorf("failed to remove container root directory %q: %v",
+			containerRootDir, err)
+	}
+
+	// Delete container metadata.
+	if err := c.containerStore.Delete(id); err != nil {
+		return nil, fmt.Errorf("failed to delete container metadata for %q: %v", id, err)
+	}
+
+	return &runtime.RemoveContainerResponse{}, nil
+}
+
+// setContainerRemoving sets the container into removing state. In removing state, the
+// container will not be started or removed again.
```

Comment:
There is no race between `Stop` and `Remove`, because `Stop` could only happen when container is in running state; `Remove` could only happen when container is in created/exited state.

For exited container, it will not be running again, so it's fine;
For created container, because `removing` avoid other operation starting the container again, so it will not go into running state.